controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -88,64 +95,67 @@ public SegmentHelper(final ConnectionFactory clientCF, HostControllerStore hostS
         final String qualifiedStreamSegmentName = getQualifiedStreamSegmentName(scope, stream, segmentId);
         final Controller.NodeUri uri = getSegmentUri(scope, stream, segmentId);
-        final WireCommandType type = WireCommandType.CREATE_SEGMENT;
-        final long requestId = (clientRequestId == RequestTag.NON_EXISTENT_ID) ? idGenerator.get() : clientRequestId;
-
-        final FailingReplyProcessor replyProcessor = new FailingReplyProcessor() {
-            @Override
-            public void connectionDropped() {
-                log.warn(requestId, "CreateSegment {} Connection dropped", qualifiedStreamSegmentName);
-                result.completeExceptionally(
-                        new WireCommandFailedException(type, WireCommandFailedException.Reason.ConnectionDropped));
-            }
-
-            @Override
-            public void wrongHost(WireCommands.WrongHost wrongHost) {
-                log.warn(requestId, "CreateSegment {} wrong host", qualifiedStreamSegmentName);
-                result.completeExceptionally(new WireCommandFailedException(type, WireCommandFailedException.Reason.UnknownHost));
-            }
-
-            @Override
-            public void segmentAlreadyExists(WireCommands.SegmentAlreadyExists segmentAlreadyExists) {
-                log.info(requestId, "CreateSegment {} segmentAlreadyExists", qualifiedStreamSegmentName);
-                result.complete(true);
-            }
-
-            @Override
-            public void segmentCreated(WireCommands.SegmentCreated segmentCreated) {
-                log.info(requestId, "CreateSegment {} SegmentCreated", qualifiedStreamSegmentName);
-                result.complete(true);
-            }
-
-            @Override
-            public void processingFailure(Exception error) {
-                log.error(requestId, "CreateSegment {} threw exception", qualifiedStreamSegmentName, error);
-                handleError(error, result, type);
-            }
-
-            @Override
-            public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenCheckFailed) {
-                result.completeExceptionally(
-                        new WireCommandFailedException(new AuthenticationException(authTokenCheckFailed.toString()),
-                                type, WireCommandFailedException.Reason.AuthFailed));
-            }
-        };
 
+        RawClient connection = new RawClient(ModelHelper.encode(uri), connectionFactory);
+        final long requestId = connection.getFlow().getNextSequenceNumber();
         Pair<Byte, Integer> extracted = extractFromPolicy(policy);
 
-        WireCommands.CreateSegment request = new WireCommands.CreateSegment(requestId, qualifiedStreamSegmentName,
-                extracted.getLeft(), extracted.getRight(), controllerToken);
-        sendRequestAsync(request, replyProcessor, result, ModelHelper.encode(uri));
-        return result;
+        Class[] expectedArray = {WireCommands.SegmentCreated.class,
```

Comment:
I think this should be called 'expectedReplies'. It's not helpful to put the type into the name, but it's actually confusing because it implies that you are expecting all of them.

Reply:
Do we want to have a map of expected response and expected completion result.. also possible logging. 
that will keep parity with existing implementation. 

in some specific cases, we may want to fail request will a specific error. 
this implementation will limit that possibility. but having a map will keep those possibilities open.